arduino-ide-extension/src/browser/contributions/save-as-sketch.ts

```diff
@@ -84,91 +142,157 @@ export class SaveAsSketch extends SketchContribution {
+    while (!sketchFolderDestinationUri) {
+      const { filePath } = await remote.dialog.showSaveDialog(
+        remote.getCurrentWindow(),
+        {
+          title: nls.localize(
+            'arduino/sketch/saveFolderAs',
+            'Save sketch folder as...'
+          ),
+          defaultPath,
+        }
+      );
+      if (!filePath) {
+        return undefined;
       }
-    );
-    if (!filePath || canceled) {
-      return false;
-    }
-    const destinationUri = await this.fileSystemExt.getUri(filePath);
-    if (!destinationUri) {
-      return false;
-    }
-    const workspaceUri = await this.sketchService.copy(sketch, {
-      destinationUri,
-    });
-    if (workspaceUri) {
-      await this.saveOntoCopiedSketch(sketch.mainFileUri, sketch.uri, workspaceUri);
-      if (markAsRecentlyOpened) {
-        this.sketchService.markAsRecentlyOpened(workspaceUri);
+      const destinationUri = await this.fileSystemExt.getUri(filePath);
+      // The new location of the sketch cannot be inside the location of current sketch.
+      // https://github.com/arduino/arduino-ide/issues/1882
+      let dialogContent: InvalidSketchFolderDialogContent | undefined;
+      if (new URI(sketch.uri).isEqualOrParent(new URI(destinationUri))) {
+        dialogContent = {
+          message: nls.localize(
+            'arduino/sketch/invalidSketchFolderLocationMessage',
+            "Invalid sketch folder location: '{0}'",
+            filePath
+          ),
+          details: nls.localize(
+            'arduino/sketch/invalidSketchFolderLocationDetails',
+            'You cannot save a sketch into a folder inside itself. This would go on forever.'
+          ),
+          question: nls.localize(
+            'arduino/sketch/editInvalidSketchFolderLocationQuestion',
+            'Do you want to try to save the sketch to a different location?'
+          ),
+        };
       }
-    }
-    const options: WorkspaceInput & StartupTask.Owner = {
-      preserveWindow: true,
-      tasks: [],
-    };
-    if (workspaceUri && openAfterMove) {
-      this.windowService.setSafeToShutDown();
-      if (wipeOriginal || (openAfterMove && execOnlyIfTemp)) {
-        options.tasks.push({
-          command: DeleteSketch.Commands.DELETE_SKETCH.id,
-          args: [sketch.uri],
-        });
+      if (!dialogContent) {
+        const sketchFolderName = new URI(destinationUri).path.base;
+        const errorMessage = Sketch.validateSketchFolderName(sketchFolderName);
+        if (errorMessage) {
+          dialogContent = {
+            message: nls.localize(
+              'arduino/sketch/invalidSketchFolderNameMessage',
+              "Invalid sketch folder name: '{0}'",
+              sketchFolderName
+            ),
+            details: errorMessage,
+            question: nls.localize(
+              'arduino/sketch/editInvalidSketchFolderQuestion',
+              'Do you want to try to save the sketch with a different name?'
```

Comment:
```suggestion
              'Do you want to try saving the sketch with a different name?'
```

See my explanation in the suggestion above.